### PR TITLE
Added Cpulimit to Photon OS

### PR DESCRIPTION
**Pull request by bkwzwz**
Added cpulimit utility to Photon OS. This utility can help to limit the CPU of a process in a kubernetes environment.  Without this, a high CPU process such as tar with gz can cause kubernetes to detect CPU running at 100% and spawn another container to share the workload.  With this utility, we can limit the CPU to a preset percentage which will not trigger the spawning.

### CODE REVIEW

File: SPECS/cpulimit/cpulimit.spec [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+Version:        1.1	
+Release:	1%{?dist}
+Summary:	CPU Usage Limiter for Linux
+
+Group:		Applications/System
+License:	GPLv2+
+URL:		http://cpulimit.sourceforge.net/
+Source0:	http://downloads.sourceforge.net/project/%{name}/%{name}/%{name}/%{name}-%{version}.tar.gz
+BuildRoot:	%(mktemp -ud %{_tmppath}/%{name}-%{version}-%{release}-XXXXXX)
+
```
Comment (tapakund) [inline SPECS/cpulimit/cpulimit.spec:11]: Please add:
Vendor:         VMware, Inc.
Distribution:   Photon

File: SPECS/cpulimit/cpulimit.spec [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+%description
+cpulimit is a simple program which attempts to limit the CPU usage of a process
+(expressed in percentage, not in CPU time). This is useful to control batch
+jobs, when you don't want them to eat too much CPU. It does not act on the nice
+value or other scheduling priority stuff, but on the real CPU usage. Also, it
+is able to adapt itself to the overall system load, dynamically and quickly.
+
+%prep
+%setup -q
```
Comment (tapakund) [inline SPECS/cpulimit/cpulimit.spec:20]: use autosetup

File: SPECS/cpulimit/cpulimit.spec [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+
+%build
+gcc $RPM_OPT_FLAGS -lrt -o cpulimit cpulimit.c
+
+
+%install
+rm -rf $RPM_BUILD_ROOT
```
Comment (tapakund) [inline SPECS/cpulimit/cpulimit.spec:28]: use %{buildroot}

File: SPECS/cpulimit/cpulimit.spec [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+install -Dp -m 755 cpulimit $RPM_BUILD_ROOT/%{_bindir}/cpulimit
+
+%clean
+rm -rf $RPM_BUILD_ROOT
+
+%files
+%{_bindir}/cpulimit
```
Comment (tapakund) [inline SPECS/cpulimit/cpulimit.spec:35]: add %defattr(-,root,root)

File: SPECS/cpulimit/cpulimit.spec [hunk 5]
```diff
@@ -0,0 +1,39 @@
+Name:		cpulimit
+Version:        1.1	
+Release:	1%{?dist}
+Summary:	CPU Usage Limiter for Linux
+
+Group:		Applications/System
+License:	GPLv2+
+URL:		http://cpulimit.sourceforge.net/
+Source0:	http://downloads.sourceforge.net/project/%{name}/%{name}/%{name}/%{name}-%{version}.tar.gz
+BuildRoot:	%(mktemp -ud %{_tmppath}/%{name}-%{version}-%{release}-XXXXXX)
```
Comment (tapakund) [inline SPECS/cpulimit/cpulimit.spec:10]: Why do we need to set BuildRoot:	%(mktemp -ud %{_tmppath}/%{name}-%{version}-%{release}-XXXXXX) ?

File: SPECS/cpulimit/cpulimit.spec [hunk 6]
```diff
@@ -0,0 +1,39 @@
+Name:		cpulimit
+Version:        1.1	
+Release:	1%{?dist}
+Summary:	CPU Usage Limiter for Linux
+
+Group:		Applications/System
+License:	GPLv2+
+URL:		http://cpulimit.sourceforge.net/
+Source0:	http://downloads.sourceforge.net/project/%{name}/%{name}/%{name}/%{name}-%{version}.tar.gz
```
Comment (tapakund) [inline SPECS/cpulimit/cpulimit.spec:9]: also please add sha sum for the source.

**Review by tapakund — COMMENTED** [review]
some nits and comments.

File: SPECS/cpulimit/cpulimit.spec [hunk 7]
```diff
@@ -0,0 +1,39 @@
+Name:		cpulimit
```
Comment (sshedi) [inline SPECS/cpulimit/cpulimit.spec:1]: nit: fix indentation here. refer other specs

Also, add:
```
BuildRequires: glibc-devel

Requires: glibc
```

File: SPECS/cpulimit/cpulimit.spec [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+jobs, when you don't want them to eat too much CPU. It does not act on the nice
+value or other scheduling priority stuff, but on the real CPU usage. Also, it
+is able to adapt itself to the overall system load, dynamically and quickly.
+
+%prep
+%setup -q
+
+
+%build
+gcc $RPM_OPT_FLAGS -lrt -o cpulimit cpulimit.c
```
Comment (sshedi) [inline SPECS/cpulimit/cpulimit.spec:24]: use make %{?_smp_mflags}

File: SPECS/cpulimit/cpulimit.spec [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+
+%install
+rm -rf $RPM_BUILD_ROOT
+install -Dp -m 755 cpulimit $RPM_BUILD_ROOT/%{_bindir}/cpulimit
```
Comment (sshedi) [inline SPECS/cpulimit/cpulimit.spec:29]: nit: Will be good to use
`install -Dp -m 755 %{name} %{buildroot}/%{_bindir}/%{name}`

**Review by sshedi — CHANGES_REQUESTED** [review]
If possible, please provide a simple script to test this binary(just attach it to discussions here)
we will use it in our internal test repos.

**Comment by bkwzwz** [comment]
> Also, do we really need this?
> have you explored on https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html

Yes, we need this.  For k8s/container, we don't have systemd running.

File: SPECS/cpulimit/cpulimit.spec (hunk 2 above)
Comment (bkwzwz) [reply SPECS/cpulimit/cpulimit.spec]: Fixed.

File: SPECS/cpulimit/cpulimit.spec (hunk 4 above)
Comment (bkwzwz) [reply SPECS/cpulimit/cpulimit.spec:39]: Added.

File: SPECS/cpulimit/cpulimit.spec (hunk 6 above)
Comment (bkwzwz) [reply SPECS/cpulimit/cpulimit.spec]: Added.

File: SPECS/cpulimit/cpulimit.spec (hunk 7 above)
Comment (bkwzwz) [reply SPECS/cpulimit/cpulimit.spec]: Added.

File: SPECS/cpulimit/cpulimit.spec (hunk 8 above)
Comment (bkwzwz) [reply SPECS/cpulimit/cpulimit.spec:27]: I am not sure about this.  I copied this spec from another spec which has this.

File: SPECS/cpulimit/cpulimit.spec (hunk 9 above)
Comment (bkwzwz) [reply SPECS/cpulimit/cpulimit.spec]: Fixed.

**Review by sshedi — CHANGES_REQUESTED** [review]

File: SPECS/cpulimit/cpulimit.spec (hunk 8 above)
Comment (bkwzwz) [reply SPECS/cpulimit/cpulimit.spec]: I see it now. I agree it is better to use 'make'.  Switched to use 'make' instead.

**Comment by sshedi** [comment]
Can you please explain what your idea here is? How are you going to use cpulimit binary?

Please refer:
https://www.thorsten-hans.com/docker-container-cpu-limits-explained/
https://kubernetes.io/docs/tasks/configure-pod-container/assign-cpu-resource/

Just want to be on the same page as you.

**Comment by sshedi** [comment]
Inside container you can't control cpu usage, there is no cgroup inside container. I'm saying this because you have mentioned `"For k8s/container, we don't have systemd running."`

Are you planning to use cpulimit inside container? Have you tried it?

**Comment by bkwzwz** [comment]
> Can you please explain what your idea here is? How are you going to use cpulimit binary?
> 
> Please refer: https://www.thorsten-hans.com/docker-container-cpu-limits-explained/ https://kubernetes.io/docs/tasks/configure-pod-container/assign-cpu-resource/
> 
> Just want to be on the same page as you.

Hi, I read the above and it talks about how to limit a container CPU from host perspective.  What we need is to limit a process  within a container to not to exceed a threshold.  

In k8s, we have set:

targetCPUUtilizationPercentage=50

So when k8s has detected the high CPU usage of our service running, it can scale the number of pods to accommodate
the load.  For example, if a pod is at 100%, it will launch 3 pods where each will have 33% CPU load.

However, during maintenance time, we may run:

tar -czvf example.tar.gz $folder

For 60 sec, the CPU will go to 100% and accidentally trigger k8s to launch a new pod.

Therefore, we can use cpulimit to control the CPU to 25% like below:

cpulimit -l 25 -i tar -czvf example.tar.gz $folder

When cpulimit above, we can maintain the pod CPU to below 50% threshold and not triggering another pod.

**Comment by sshedi** [comment]
Changes have been merged:
https://github.com/vmware/photon/commit/87d2d8982b5956dfb301f063e3795425ee42bfd0

@suezzelur please close this PR.

**Pull request closed without merging**